### ISSUE TITLE
How does the policy gradient loss improve the model when training accuracy = 100%

### ISSUE THREAD

**Issue by Qinyu-Allen-Zhao** [ypwang61/One-Shot-RLVR#1]
Hi Team,

Thank you for your interesting and insightful work! 

The paper shows the phenomenon of post-saturation generalization: the model’s training accuracy on the single example is already saturated, but the test performance continues improving. And the policy gradient loss is the main contributor for the model performance.

But I'm wondering how policy gradient loss keeps improving the model after the training accuracy achieves 100%. 

Assume that, after K training steps, the model can always solve the single training sample. The rewards are all 1s in the group, that means, the advantages are 0s in GRPO. The policy gradient loss should be 0. Is this correct? Could you provide more insights on this? Thank you!

Qinyu

**Comment by ypwang61** [comment]
Hi Qinyu

Thanks for your interest! Yeah this is a great question that confuses us before. I think there are several reasons:

1. A lot of improvement are get before the training accuracy comes to 100%, for example, before the first 100 steps
2. Another most important one, even if the training acc gets to near 100%, it's not always 100% (for example, 99.xx%), and due to the calculation of RL algorithm (which amplify the advantage of outliers), these responses can still provide some policy gradient. Especially when we add entropy loss to motivate model to explore more, there will be more failure rollouts, and these policy gradient become actually unignorable. When we remove entropy loss, policy gradient loss become sparser after 500 steps (but still not zero due to random rollout since the non-zero training temperature), match our conclusion from ablation study that encouraging model to explore more is sometimes helpful.


You can see the `actor/pg_loss` in this wandb record: [https://wandb.ai/yipingwanguw/verl_few_shot?nw=nwuseryipingwang22
](https://wandb.ai/yipingwanguw/verl_few_shot?nw=nwuseryipingwang22), which shows non-zero policy gradient loss

**Comment by Qinyu-Allen-Zhao** [comment]
Thank you for your detailed response!
